File: ruv-swarm/npm/src/onboarding/claude-detector.js
/**
 * Claude Code Detection Module
 * Detects Claude Code installation across different platforms
 */

import { execSync } from 'child_process';
import { existsSync } from 'fs';
import path from 'path';
import os from 'os';

/**
 * Detects if Claude Code is installed and returns installation info
 * @returns {Promise<{installed: boolean, path?: string, version?: string}>}
 */
export async function detectClaudeCode() {
  const platform = os.platform();
  
  try {
    // Try to find Claude Code using 'which' command (Unix) or 'where' (Windows)
    const command = platform === 'win32' ? 'where' : 'which';
    const claudePath = execSync(`${command} claude-code`, { encoding: 'utf8' }).trim();
    
    if (claudePath) {
      // Get version
      const version = await getClaudeVersion(claudePath);
      return {
        installed: true,
        path: claudePath,
        version
      };
    }
  } catch (error) {
    // Command not found in PATH, check common installation locations
  }

  // Check platform-specific common locations
  const commonPaths = getCommonPaths(platform);
  
  for (const checkPath of commonPaths) {
    if (existsSync(checkPath)) {
      const version = await getClaudeVersion(checkPath);
      return {
        installed: true,
        path: checkPath,
        version
      };
    }
  }

  return { installed: false };
}

/**
 * Get Claude Code version from binary
 * @param {string} claudePath - Path to Claude Code binary
 * @returns {Promise<string|null>}
 */
async function getClaudeVersion(claudePath) {
  try {
    const versionOutput = execSync(`"${claudePath}" --version`, { encoding: 'utf8' });
    // Extract version from output (format: "Claude Code vX.Y.Z")
    const match = versionOutput.match(/v?(\d+\.\d+\.\d+)/);
    return match ? match[1] : null;
  } catch (error) {
    return null;
  }
}

/**
 * Get common installation paths by platform
 * @param {string} platform - Node.js platform string
 * @returns {string[]}
 */
function getCommonPaths(platform) {
  const homeDir = os.homedir();
  
  switch (platform) {
    case 'win32':
      return [
        path.join(process.env.LOCALAPPDATA || '', 'Programs', 'claude-code', 'claude-code.exe'),
        path.join(process.env.PROGRAMFILES || '', 'Claude Code', 'claude-code.exe'),
        path.join(homeDir, 'AppData', 'Local', 'Programs', 'claude-code', 'claude-code.exe'),
        path.join(homeDir, '.local', 'bin', 'claude-code.exe')
      ];
    
    case 'darwin':
      return [
        '/Applications/Claude Code.app/Contents/MacOS/claude-code',
        path.join(homeDir, 'Applications', 'Claude Code.app', 'Contents', 'MacOS', 'claude-code'),
        '/usr/local/bin/claude-code',
        path.join(homeDir, '.local', 'bin', 'claude-code')
      ];
    
    default: // Linux and other Unix-like
      return [
        '/usr/local/bin/claude-code',
        '/usr/bin/claude-code',
        '/opt/claude-code/claude-code',
        path.join(homeDir, '.local', 'bin', 'claude-code'),
        path.join(homeDir, 'bin', 'claude-code')
      ];
  }
}

/**
 * Check if Claude Code version is compatible
 * @param {string} version - Version string
 * @returns {boolean}
 */
export function isVersionCompatible(version) {
  if (!version) return false;
  
  const [major, minor] = version.split('.').map(Number);
  // Assuming we need at least version 1.0.0
  return major >= 1;
}

export default {
  detectClaudeCode,
  isVersionCompatible
};

File: ruv-swarm/crates/ruv-swarm-cli/src/onboarding/claude_detector.rs
//! Claude Code detection for cross-platform onboarding

use anyhow::{anyhow, Result};
use async_trait::async_trait;
use std::path::{Path, PathBuf};
use std::process::Command;

/// Information about Claude Code installation
#[derive(Debug, Clone)]
pub struct ClaudeInfo {
    pub installed: bool,
    pub path: String,
    pub version: Option<String>,
}

/// Trait for Claude Code detection
#[async_trait]
pub trait ClaudeDetector {
    async fn detect(&self) -> Result<ClaudeInfo>;
    async fn validate_installation(&self, path: &str) -> Result<bool>;
}

/// Default implementation of Claude Code detector
pub struct DefaultClaudeDetector {
    search_paths: Vec<PathBuf>,
}

impl DefaultClaudeDetector {
    pub fn new() -> Self {
        Self {
            search_paths: Self::get_default_search_paths(),
        }
    }
    
    /// Get platform-specific search paths for Claude Code
    fn get_default_search_paths() -> Vec<PathBuf> {
        let mut paths = Vec::new();
        
        // Common paths where Claude Code might be installed
        if cfg!(target_os = "windows") {
            // Windows paths
            paths.extend(vec![
                PathBuf::from("C:\\Program Files\\Claude\\claude.exe"),
                PathBuf::from("C:\\Program Files (x86)\\Claude\\claude.exe"),
                PathBuf::from(format!("{}\\AppData\\Local\\Claude\\claude.exe", 
                    std::env::var("USERPROFILE").unwrap_or_default())),
            ]);
        } else if cfg!(target_os = "macos") {
            // macOS paths
            paths.extend(vec![
                PathBuf::from("/Applications/Claude.app/Contents/MacOS/claude"),
                PathBuf::from(format!("{}/Applications/Claude.app/Contents/MacOS/claude",
                    std::env::var("HOME").unwrap_or_default())),
            ]);
        } else {
            // Linux paths
            paths.extend(vec![
                PathBuf::from("/usr/local/bin/claude"),
                PathBuf::from("/usr/bin/claude"),
                PathBuf::from(format!("{}/.local/bin/claude",
                    std::env::var("HOME").unwrap_or_default())),
                PathBuf::from("/opt/claude/claude"),
            ]);
        }
        
        // Also check PATH
        if let Ok(path_env) = std::env::var("PATH") {
            for path_dir in std::env::split_paths(&path_env) {
                let claude_path = if cfg!(target_os = "windows") {
                    path_dir.join("claude.exe")
                } else {
                    path_dir.join("claude")
                };
                paths.push(claude_path);
            }
        }
        
        paths
    }
    
    /// Detect Claude Code version
    async fn get_version(&self, path: &str) -> Option<String> {
        let output = Command::new(path)
            .arg("--version")
            .output()
            .ok()?;
            
        if output.status.success() {
            String::from_utf8(output.stdout)
                .ok()
                .map(|s| s.trim().to_string())
        } else {
            None
        }
    }
}

#[async_trait]
impl ClaudeDetector for DefaultClaudeDetector {
    async fn detect(&self) -> Result<ClaudeInfo> {
        // First try to find Claude in the search paths
        for path in &self.search_paths {
            if path.exists() {
                let path_str = path.to_string_lossy().to_string();
                if self.validate_installation(&path_str).await? {
                    let version = self.get_version(&path_str).await;
                    return Ok(ClaudeInfo {
                        installed: true,
                        path: path_str,
                        version,
                    });
                }
            }
        }
        
        // Try to find Claude using `which` or `where` command
        let find_cmd = if cfg!(target_os = "windows") { "where" } else { "which" };
        let claude_name = if cfg!(target_os = "windows") { "claude.exe" } else { "claude" };
        
        if let Ok(output) = Command::new(find_cmd).arg(claude_name).output() {
            if output.status.success() {
                let path_str = String::from_utf8(output.stdout)
                    .map_err(|_| anyhow!("Invalid UTF-8 in command output"))?
                    .trim()
                    .to_string();
                    
                if self.validate_installation(&path_str).await? {
                    let version = self.get_version(&path_str).await;
                    return Ok(ClaudeInfo {
                        installed: true,
                        path: path_str,
                        version,
                    });
                }
            }
        }
        
        // Claude Code not found
        Ok(ClaudeInfo {
            installed: false,
            path: String::new(),
            version: None,
        })
    }
    
    async fn validate_installation(&self, path: &str) -> Result<bool> {
        if !Path::new(path).exists() {
            return Ok(false);
        }
        
        // Try to run Claude with --help to validate it's working
        let output = Command::new(path)
            .arg("--help")
            .output()
            .map_err(|_| anyhow!("Failed to execute Claude command"))?;
            
        Ok(output.status.success())
    }
}

impl Default for DefaultClaudeDetector {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[tokio::test]
    async fn test_claude_detector_creation() {
        let detector = DefaultClaudeDetector::new();
        assert!(!detector.search_paths.is_empty());
    }
    
    #[tokio::test]
    async fn test_invalid_path_validation() {
        let detector = DefaultClaudeDetector::new();
        let result = detector.validate_installation("/nonexistent/path").await;
        assert!(result.is_ok());
        assert!(!result.unwrap());
    }
}